mesonbuild/ast/introspection.py

```diff
@@ -15,47 +15,56 @@
 
-from . import AstInterpreter
+from .interpreter import AstInterpreter
+from .visitor import AstVisitor
 from .. import compilers, environment, mesonlib, optinterpreter
 from .. import coredata as cdata
 from ..mesonlib import MachineChoice
-from ..interpreterbase import InvalidArguments
+from ..interpreterbase import InvalidArguments, TYPE_nvar
 from ..build import Executable, Jar, SharedLibrary, SharedModule, StaticLibrary
 from ..mparser import BaseNode, ArithmeticNode, ArrayNode, ElementaryNode, IdNode, FunctionNode, StringNode
+from typing import Any, Dict, List, Optional
 import os
 
 build_target_functions = ['executable', 'jar', 'library', 'shared_library', 'shared_module', 'static_library', 'both_libraries']
 
 class IntrospectionHelper:
     # mimic an argparse namespace
-    def __init__(self, cross_file):
-        self.cross_file = cross_file
-        self.native_file = None
-        self.cmd_line_options = {}
+    def __init__(self, cross_file: str) -> None:
+        self.cross_file = cross_file  # type: str
+        self.native_file = None       # type: str
+        self.cmd_line_options = {}    # type: Dict[str, str]
 
 class IntrospectionInterpreter(AstInterpreter):
     # Interpreter to detect the options without a build directory
     # Most of the code is stolen from interpreter.Interpreter
-    def __init__(self, source_root, subdir, backend, visitors=None, cross_file=None, subproject='', subproject_dir='subprojects', env=None):
+    def __init__(self,
+                 source_root: str,
+                 subdir: str,
+                 backend: str,
+                 visitors: Optional[List[AstVisitor]] = None,
+                 cross_file: Optional[str] = None,
+                 subproject: str = '',
+                 subproject_dir: str = 'subprojects',
+                 env: Optional[environment.Environment] = None) -> None:
         visitors = visitors if visitors is not None else []
-        super().__init__(source_root, subdir, visitors=visitors)
+        super().__init__(source_root, subdir, subproject, visitors=visitors)
 
         options = IntrospectionHelper(cross_file)
         self.cross_file = cross_file
         if env is None:
             self.environment = environment.Environment(source_root, None, options)
         else:
             self.environment = env
-        self.subproject = subproject
         self.subproject_dir = subproject_dir
         self.coredata = self.environment.get_coredata()
         self.option_file = os.path.join(self.source_root, self.subdir, 'meson_options.txt')
         self.backend = backend
         self.default_options = {'backend': self.backend}
-        self.project_data = {}
-        self.targets = []
-        self.dependencies = []
-        self.project_node = None
+        self.project_data = {}    # type: Dict[str, Any]
+        self.targets = []         # type: List[Dict[str, Any]]
+        self.dependencies = []    # type: List[Dict[str, Any]]
```

Comment:
I could theoretically replace the `Any` with a complex Union (something like `Union[str, int, bool, list, dict]`, but then there still would be a plain `list` and `dict` left unless I can somehow recursively define the type.

Also, there are no actual target classes stored here, just JSON serializable stuff. I plan to refactor this in a followup PR, so that I have an explicit, custom class for each field.

Reply:
I don't want to do this in this PR, since it is already quite bloated.